### ISSUE TITLE
In gRPC client, `ctx.log().whenComplete().thenAccept()` callback not invoked when `ConcurrencyLimitingClient` decorator is used 

### ISSUE THREAD

**Issue by Lincong** [line/armeria#4835]
I am on `1.16.2` and using gRPC client with the below decorator structure:

```scala
// Decorator 1: circuit break decorator
clientBuilder.decorator(CircuitBreakerClient.newDecorator(...))

// Decorator 2: concurrency limit decorator
val concurrencyLimit: ConcurrencyLimit = ...
val concurrencyLimitingDecorator = ConcurrencyLimitingClient.newDecorator(concurrencyLimit)
clientBuilder.decorator(concurrencyLimitingDecorator)

// Decorator 3: add a callback to `RequestLogAccess`
clientBuilder.decorator((delegate, ctx, req) => {
   // I expect this callback to always be invoked when an exception is thrown from decorator 1 or 2.
   // Please let me know if this expectation is reasonable.
   ctx.log().whenComplete().thenAccept(reqLog => { /* A callback */ })
   delegate.execute(ctx, req)
}
```

### Observed problems
1. When the circuit break decorator rejects a request (by throwing a `FailFast` exception), the callback registered in decorator 3 is not invoked.
2. When the concurrency limit decorator rejects a request (e.g. by throwing a `TooManyPendingAcquisitionsException` exception), the callback registered in decorator 3 is not invoked.

### My current workaround
I add another decorator (`decorator 2.5`) before `decorator 3` as shown below. This way, the callback in `decorator 3` is invoked when decorator 1 or 2 throws an exception.
```scala
// Decorator 1: circuit breaking decorator
// ... same as above

// Decorator 2: concurrency limit decorator
// ... same as above

// Decorator 2.5: Complete `RequestLogAccess` upon exception
clientBuilder.decorator((delegate, ctx, req) => {
  val resp = delegate.execute(ctx, req)
  resp.mapError(t => {
    // Copied from `ClientUtil.fail(ClientRequestContext ctx, Throwable cause)`
    val request: HttpRequest = ctx.request
    if (request != null) request.abort(t)
    val logBuilder: RequestLogBuilder = ctx.logBuilder
    logBuilder.endRequest(t)
    logBuilder.endResponse(t)
    t
  })
})

// Decorator 3
// ... same as above
```

### Questions
1. Are these `Observed problems` really problems?
2. Do I misunderstand the semantics of `ctx.log().whenComplete().thenAccept(...)` callback?
3. How sane is my workaround? Is there any better workaround to achieve what I want?

Thanks!

**Comment by Lincong** [comment]
Similar issue: https://github.com/line/armeria/issues/4826

**Comment by minwoox** [comment]
> Are these Observed problems really problems?

Yes, it's a problem.

> Do I misunderstand the semantics of ctx.log().whenComplete().thenAccept(...) callback?

the callback must be called in all situations. You get it right.

> How sane is my workaround? Is there any better workaround to achieve what I want?

I believe it's a good workaround. 👍

**Comment by jrhee17** [comment]
Within armeria internally, we try to ensure that both log and responses are completed.

e.g.)
https://github.com/line/armeria/blob/62349aba927cde2a8bf8936051b7346e7002a2d9/core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java#L231-L239

However, this isn't necessarily true for `ConcurrencyLimitingClient`, which is the point I think needs fixing
- code path 1: https://github.com/line/armeria/blob/62349aba927cde2a8bf8936051b7346e7002a2d9/core/src/main/java/com/linecorp/armeria/client/limit/AbstractConcurrencyLimitingClient.java#L109
- code path 2: https://github.com/line/armeria/blob/62349aba927cde2a8bf8936051b7346e7002a2d9/core/src/main/java/com/linecorp/armeria/client/limit/AbstractConcurrencyLimitingClient.java#L123-L125


> How sane is my workaround? Is there any better workaround to achieve what I want?

A lot of armeria decorators assume that `log` and `HttpResponse` are both completed, so we should also do the same in decorators like `ConcurrencyLimitingClient`. I agree with the direction of closing the `ClientRequestContext#log` when `HttpResponse` is completed with an exception.

**Comment by Lincong** [comment]
Thanks @jrhee17 @minwoox for your comments! We are on the same page. I will proceed with the `Decorator 2.5` workaround for now. Looking forward to the fix :)

**Comment by ikhoon** [comment]
This seems fixed by #4834. @jrhee17 Am I right?

**Comment by jrhee17** [comment]
I think so 😄   Thanks for checking 🙇
